{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Genes_Stacked.Bar.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "wZhxC5erkz9N"
      },
      "source": [
        "# Makes a table in which each metagenome is converted to a table of [function, number]\n",
        "# the lines of the table are written to a \"Clean\" file\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "\n",
        "search_genes = 'ab' # ab: antibiotic resistance; tx: toxins; mt: heavy metals\n",
        "\n",
        "infiles = [ '/content/gdrive/Shared drives/4.Venetian Canals/Joined MG Files/1.1.Vir.mgm4876242.3.csv', \n",
        "            '/content/gdrive/Shared drives/4.Venetian Canals/Joined MG Files/2.3.Vir.mgm4876251.3.csv',\n",
        "            '/content/gdrive/Shared drives/4.Venetian Canals/Joined MG Files/3.2.Vir.mgm4876251.3.csv',\n",
        "            '/content/gdrive/Shared drives/4.Venetian Canals/Joined MG Files/4.1.Vir.mgm4876246.3.csv',\n",
        "            '/content/gdrive/Shared drives/4.Venetian Canals/Joined MG Files/5.3.Vir.mgm4876239.3.csv',\n",
        "            '/content/gdrive/Shared drives/4.Venetian Canals/Joined MG Files/6.2.Vir.mgm4876231.3.csv'\n",
        "           ]\n",
        "labels =  ['S1', 'S2', 'S3', 'S4', 'S5', 'S6'] \n",
        "rast_totals = [196401, 102523, 117807, 63138,100590, 101290]\n",
        "\n",
        "if search_genes == 'ab':\n",
        "  title = 'Antibiotic Resistance Genes'\n",
        "  nots = 'empty'\n",
        "  targets = [ \n",
        "           ['Vancomycin'], ['Acriflavine'], ['Fosfomycin'],\n",
        "           ['Streptomycin'], ['Azithromycin'], ['Lincomycin'], \n",
        "           ['Trimethoprim'], ['Rifam'] , ['Aminoglycoside'],\n",
        "           ['Metronidazole'], ['Nitroimidazole'],  ['Quinolone'], ['Macrolide'], ['quninolon'],            \n",
        "           ['Polymyxin'], ['Topoisomerase inhibitor'], ['Beta-lactamase'], ['mycin'], ['Multidrug'] ]  #\n",
        "\n",
        "elif search_genes == 'mt':\n",
        "  targets = [ ['Cobalt'], ['Mercury'] , ['Zinc'], ['Arsenic'], ['Cadmium'], ['Copper'], \n",
        "             ['Chromate'], ['Lead'], ['Co/Zn/Cd'] ]\n",
        "  title = 'Heavy Metal Resistance Genes'\n",
        "  nots = 'precorrin'\n",
        "\n",
        "elif search_genes == 'tx': \n",
        "  targets = [ ['Toxin'], ['Hemolysin'], ['Leukocidin'], ['Collagenase'], ['Hyaluronate lyase'], \n",
        "             ['Fibronectin'], ['Type iii'],['Multiple'] ]\n",
        "  title = 'Toxin Genes'\n",
        "  nots = 'empty'\n",
        "\n",
        "# for reading files and converting each taxon to a list of [taxon, number]\n",
        "clean_lines =[]\n",
        "for x in range(len(infiles)):\n",
        "  with open(infiles[x], 'r') as f:\n",
        "    data1 = f.read()\n",
        "    data2 = data1.lower().replace('co/zn/cd', 'cobalt, zinc, cadmium').\\\n",
        "    replace('gyrase', 'topoisomerase inhibitor').replace('erythromycin', 'macrolide').\\\n",
        "    replace('acriflavin', 'acriflavine').replace('macrolide', 'macrolide').\\\n",
        "    replace('polymyxin', 'aminoglycoside').replace('topoisomerase', 'topoisomerase inhibitor').\\\n",
        "    replace('mercuric', 'mercury')\n",
        "    #print(repr(data[-100:]))\n",
        "    lines = data2.split('\\n')\n",
        "    long_lines =[]\n",
        "    for y in range(len(lines)):\n",
        "      if len(str(lines[y])) >120: \n",
        "        long_lines.append(lines[y])    \n",
        "    clean_lines.append(long_lines)\n",
        "    print('Number of genes at Site', x+1, '=', len(long_lines))\n",
        "  f.close()\n",
        "\n",
        "multiples =[]\n",
        "for x in range(len(targets)-1):\n",
        "  multiples.append(targets[x][0])\n",
        "print('\\nList for finding multiples:', multiples)\n",
        "print()\n",
        "\n",
        "# make table with each row being  a list with one target and 'numer of files' 0's\n",
        "for a in range(len(targets)):\n",
        "  for b in range(len(infiles)):\n",
        "    targets[a].append(0)\n",
        "\n",
        "funcset = set() \n",
        "dup_hits, mon_hits, total_hits = 0, 0, 0 \n",
        "for a in range(len(clean_lines)):\n",
        "  for b in range(len(clean_lines[a])):\n",
        "    for c in range(len(targets)):\n",
        "      annotation = clean_lines[a][b]  \n",
        "      if targets[c][0].lower() in annotation and annotation not in funcset and\\\n",
        "                 nots not in annotation: \n",
        "        total_hits +=1\n",
        "        funcset.add(annotation)\n",
        "        duplicate = False\n",
        "        for d in range(c+1, len(multiples)):\n",
        "          if multiples[d].lower() in annotation:\n",
        "            duplicate = True \n",
        "            break                      \n",
        "        if duplicate == False: \n",
        "          targets[c][a+1] +=1 / len(clean_lines[a])\n",
        "          mon_hits +=1\n",
        "        elif duplicate == True:  \n",
        "          targets[len(targets)-1][a+1] += 1 / len(clean_lines[a])\n",
        "          dup_hits +=1  \n",
        "\n",
        "for x in range(len(targets)):\n",
        "  targets[x].append(sum(targets[x][1:]))\n",
        "#for x in targets: print(x)\n",
        "numbars = len(targets[0])-1\n",
        "targets.sort(key=lambda x: x[numbars] )\n",
        "targets.reverse()\n",
        "#for x in targets: print(x)\n",
        "for x in range(len(targets)):\n",
        "  targets[x].pop()\n",
        "\n",
        "numfunc =0\n",
        "plotfunc =[]\n",
        "for a in range(len(targets)):\n",
        "  positives =0\n",
        "  for b in range(1,len(targets[a])):\n",
        "    if targets[a][b] >0: \n",
        "      positives +=1\n",
        "  if positives >= 0.6 *len(labels) : \n",
        "    numfunc +=1\n",
        "    plotfunc.append(targets[a])\n",
        "print()\n",
        "print('\\ntotal_hits =', total_hits, ', dup_hits =', dup_hits, ', mon_hits =', mon_hits)\n",
        "\n",
        "numsites =len(labels)\n",
        "barlines =[]\n",
        "colors = ['r', 'c', '#ffde00', 'g', 'b', '#7cd700', 'm','#400000', '#8f6d5f', '#ff0087', '#33FF90'] \n",
        "\n",
        "for x in range(len(plotfunc)):\n",
        "  for y in range(1, len(plotfunc[x])):\n",
        "    plotfunc[x][y] = float(plotfunc[x][y])\n",
        "plotname =[]\n",
        "for x in range(numfunc):\n",
        "  plotname.append(plotfunc[x][0])\n",
        "\n",
        "for x in range(numfunc):\n",
        "  barlines.append(plotfunc[x][1:])\n",
        "  print(plotfunc[x])\n",
        "width = 0.5       # the width of the bars: can also be len(x) sequence\n",
        "fig, ax = plt.subplots(dpi=300)\n",
        "ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))\n",
        "\n",
        "bottomsum = [] \n",
        "for x in range(len(labels)):\n",
        "  bottomsum.append(0)\n",
        "\n",
        "for x in range(len(plotname)):\n",
        "  if x == 0:\n",
        "    ax.bar(labels, barlines[x], width, color= colors[x], bottom= bottomsum, label=plotname[x]) \n",
        "  else:\n",
        "    ax.bar(labels, barlines[x], width, color= colors[x % len(colors)], \n",
        "           bottom=[sum(z) for z in zip(bottomsum, barlines[x-1]) ], label=plotname[x]) \n",
        "    bottomsum = [sum(z) for z in zip(bottomsum, barlines[x-1]) ] \n",
        "\n",
        "ax.set_ylabel('Frequencies', fontsize=14)\n",
        "ax.set_xlabel('Site', fontsize=14)\n",
        "ax.set_title(title, fontsize = 16)\n",
        "ax.legend(loc='center left', bbox_to_anchor=(1, 0.5), fontsize=12)\n",
        "plt.show()\n",
        "print('\\nDone.')"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}